### ISSUE TITLE
Messaging getToken() always returns initialToken

### ISSUE THREAD

**Issue by britianoates** [invertase/react-native-firebase#1510]
### Issue

messaging.getToken() keeps returning the first token, even if the firebaseId has been deleted and a new token generated.

To Reproduce.  In a onTokenRefresh callback, log out the parameter token and a messaging.getToken and compare.  No matter how many times you delete the instanceId generating new tokens you'll always receive the first token.

### Environment

1. Application Target Platform:

Both

2. Development Operating System:

mac High Sierra

3. Build Tools:

xcode 9.4, Android Studio 28

4. `React Native` version:

    "react-native": "0.55.4",

5. `React Native Firebase` Version:

    "react-native-firebase": "^4.0.3",

6. `Firebase` Module:

firebase-core:12.0.1
firebase-analytics:12.0.1
firebase-messaging:12.0.1

7. Are you using `typescript`?

no

**Comment by britianoates** [comment]
https://github.com/invertase/react-native-firebase/blob/a506d1e5b5c8293630ecd0c8bcdc9e42dcf7e7c2/ios/RNFirebase/messaging/RNFirebaseMessaging.m#L111

**Comment by smathson** [comment]
@britianoates If I read this correctly it would only impact the case where the token changed during a single app instance, correct? Like if you manually delete the firebaseId for example. Is that something you do when a user signs out?

We are trying to chase down an issue on iOS where an FCMToken seems to stop working over the course of a couple days and I'm trying to determine if token refresh has anything to do with it. According to the Firebase docs a token refresh on it's own should be a rare event.

**Comment by britianoates** [comment]
We had many users not seeing the notifications we were sending.  It was difficult to track down what percentage or when notifications stopped arriving.  What you are seeing sounds similar to what we just dealt with in prod.  To fix our users we've released a version of the app that calls RequestPermission every app opening to make sure the APNS and fcmToken remain synced.

Calling to delete the firebase instanceId has been the only way we were able to recreate the failure to receive notifications.  We do not delete the token as a part of normal app usage.

**Comment by smathson** [comment]
Interesting. Does it seem like that fixed the issue for you guys?

That sounds very similar to what we're seeing. We didn't notice this problem before our last release and in that release we added code to try and capture analytics events for how many users were denying notification permissions that had an early return if they had already enabled permissions. We also only request permission after a user signs in for the first time. Previously we had been calling requestPermission on every authenticated app opening.

**Comment by britianoates** [comment]
We have not been in production long enough to be confident our fix covers all use cases. It did cover all scenarios we were able to fake using deleteToken on a dev screen, and the one device from prod we had access to.

Sounds like your solution was the opposite of ours.  We only registered during the first run, and then the APNS and fcmToken seems to grow stale and we didn't have a mechanism to resync them.  Our fix calls requestPermission on app start (Does not open modal, but does trigger a registration method. See my other ticket.)

**Comment by jcharbo** [comment]
I'm seeing the same problem.  Users that have been receiving notifications for weeks suddenly stop receiving them.  I can have them open the app, get the token returned by firebase, put that token into the firebase console and send a message.  The console reports success but the user never receives the push.  It appears that somehow the token is invalid yet firebase doesn't see it as invalid.  This is a very serious problem as we have no way of knowing which users simply never receive their push notifications.

**Comment by smathson** [comment]
@jcharbo Are you also attempting to only call Firebase.requestPermissions() on a limited basis in your app?

**Comment by duongntb94** [comment]
I have the same issue as britianoates.  messenging().getToken() only returns the initialToken. It only happens on iOS. On Android, it works fine, when the token change, messaging().getToken() return the new token.

**Comment by BrendonSled** [comment]
This can be really easily reproduced. On iOS:
```
const oldToken = await firebase.messaging().getToken();
await firebase.messaging().deleteToken();
const newToken = await firebase.messaging().getToken();
if (oldToken === newToken) {
    console.error('Token has not been refreshed');
} 
```

**Comment by kholiavko-roman** [comment]
So, guys, what is the solution of the problem ? Just remove `if (initialToken) { `
https://github.com/invertase/react-native-firebase/blob/a506d1e5b5c8293630ecd0c8bcdc9e42dcf7e7c2/ios/RNFirebase/messaging/RNFirebaseMessaging.m#L111
as @britianoates noted ?  Is this right ?

**Comment by mskv** [comment]
Has anyone experimented with adjusting the implementation of `RNFirebaseMessaging.m`?

Is this issue on the roadmap anytime soon? Is some more reproduction instructions or experimenting needed from the library users?

**Comment by lkwjohn** [comment]
> This can be really easily reproduced. On iOS:
> 
> ```
> const oldToken = await firebase.messaging().getToken();
> await firebase.messaging().deleteToken();
> const newToken = await firebase.messaging().getToken();
> if (oldToken === newToken) {
>     console.error('Token has not been refreshed');
> } 
> ```

How does the firebase.messaging().deleteToken() work? I'm getting the error of firebase.messaging().deleteToken() is unsupported by the native Firebase SDKs

**Comment by henrymoulton** [comment]
Can confirm we're seeing this behaviour:
>I can have them open the app, get the token returned by firebase, put that token into the firebase console and send a message. The console reports success but the user never receives the push. It appears that somehow the token is invalid yet firebase doesn't see it as invalid. This is a very serious problem as we have no way of knowing which users simply never receive their push notifications.

**Comment by liokm** [comment]
Any news on this? I am facing similar issue on iOS where a freshly retrieved FCM token is claimed to be not registered from FCM server when sending notification.

**Comment by kholiavko-roman** [comment]
@liokm Are you get this error only after removing token and getting new one ? 
I also have the same problem right now in production.

**Comment by henrymoulton** [comment]
@liokm this bug report was for `"react-native-firebase": "^4.0.3",` I didn't realise how behind the dependencies of the project I'm on were so trying out the latest 5.x.x release now. Will report back.

**Comment by liokm** [comment]
@kholiavko-roman Actually my issue was a mix of two problems, one was that my app was transferred to another account and thus the team id and APNS key were invalid in FCM console and needed to be updated, otherwise "not registered" error is reported. Another problem is still waiting for feedback of logs. Have you tried to always requestPermission at start time anyway?

**Comment by kholiavko-roman** [comment]
@liokm Yes, I do `requestPermission` on each time when app starting. And also I have `onTokenRefreshListener` in my app.
With my FCM key also all is okay, I rechecked this things.

**Comment by lisichka999** [comment]
@kholiavko-roman, how did you implement onTokenRefreshListener in you App? I did not find any API for this in react-native-firebase package. Thanks

**Comment by kholiavko-roman** [comment]
@lisichka999 
It calls `onTokenRefresh` 
```
componentDidMount() {
 this.onRefreshToken = firebase
      .messaging()
      .onTokenRefresh(this.props.setDeviceId);
}

componentWillUnmount() {
    if (this.onRefreshToken) {
      this.onRefreshToken();
    }
  }
```

**Comment by Nerogee** [comment]
I'm having the same issue here. firebase 5.4.0. platform: iOS
I managed to get a new token by deleting the old one. I called **firebase.messaging().deleteToken()** before **firebase.messaging().getToken();**. And it did return me a fresh token. But that token turns out to be invalid and give me error **NotRegistered** during pushing notification. Any solution?

**Comment by kashmiry** [comment]
I am having the same issue
on RNFB 6.2.0 iOS
I can get the first token, but after deleting the token I keep getting the initial token.

**Comment by mtieltjes** [comment]
Leaving these references for anyone experiencing the same issue:
https://github.com/invertase/react-native-firebase/issues/3714
https://github.com/firebase/firebase-ios-sdk/issues/6433

**Comment by Stevemoretz** [comment]
What happened at the end???
I still can't get a new token on ios.

**Comment by britianoates** [comment]
It's been awhile. I ended up working around this by saving the token after an earlier call on my side rather than calling for it since this is broken.

**Comment by Stevemoretz** [comment]
> It's been awhile. I ended up working around this by saving the token after an earlier call on my side rather than calling for it since this is broken.

Thanks for the reply.
I'm sorry could you explain more?
Saving it on the server then what?
What if it is subscribed to a topic?
That still goes for the phone.

**Comment by Stevemoretz** [comment]
@britianoates

Would you please explain how you achieved what you're describing with some code.

**Comment by Stevemoretz** [comment]
I found the solution.

```
        messaging()
            .deleteToken(undefined,'*')
            .then(() => {
                RNRestart.Restart();

```

Use this to delete the token then restart the app with `https://www.npmjs.com/package/react-native-restart` now you're sure you get a new token :) I haven't tested this on android so I don't know if android also needs this if android doesn't have this problem just put a little if statement to do this only for ios.

**Comment by SebastianGrzymala** [comment]
It was enough for me to solve the problem (works on iOS & Android):
```
messaging().deleteToken(undefined,'*');
```

**Comment by mikehardy** [comment]
I'm going to reopen this as I think we have either a documentation issue or we need to alter the code to specify the asterisk as the default. Either way I appreciate the reports of what's working posted here, thanks

**Comment by mikehardy** [comment]
v12.x.x coming up here will have this working correct on both platforms it seems - closing this as it will be backed by an automated test in our test suite. Use the scoped token workaround documented above until on v12+ 

Thanks!

**Comment by rdsedmundo** [comment]
I'm using `v12` and `deleteToken` simply has no effect for me.

I call it with `await` and it yields `null`, then I proceed to call `getToken` again right after and the same token is returned.

Also did another test, called `deleteToken`, then tried to send a push notification to the "invalidated" token, and it unexpectedly still works.

**Comment by mikehardy** [comment]
And yet, there is an actual test I have verified (while fixing up the test, via console.log statements) appeared to work for me?

This runs every commit

https://github.com/invertase/react-native-firebase/blob/b8f80a41e864c2c8d97602df625abdd2ee6bc347/packages/messaging/e2e/messaging.e2e.js#L109-L118

**Comment by milule** [comment]
@mikehardy I have a same issue like @rdsedmundo. When i call messaging().deleteToken() and then i continue call getToken() after, the new token and previous token are same. i use "@react-native-firebase/messaging": "^12.1.0". I have test on debug and production environment

**Comment by mikehardy** [comment]
@milule can you confirm whether you were on simulator or real device? We think this actually may still be an issue from separate testing (contrary to my statement above, sorry) but only on real devices. We have not narrowed it down yet.

**Comment by edulpn** [comment]
We are getting the exact symptoms described above on Android, both simulator and real device. Both `getToken` and `onTokenRefresh` give the same previous token after `deleteToken` was called.

**Comment by mikehardy** [comment]
The failures on android are contrary to reported experience above, and my personal experience. When used properly (they are async, so you have to await...) it works fine on android we think. There still may a problem on iOS but as mentioned we have not narrowed it down. 

The best way to move this forward is to use either firebase-ios-sdk or firebase-android-sdk "quickstart" projects to reproduce this (or not) with the native SDKs so we may determine where the issue is.

https://github.com/firebase/quickstart-android/tree/master/messaging
https://github.com/firebase/quickstart-ios/tree/master/messaging

They both are ready to go out of the box I think, or maybe need your google services plist/json files but otherwise ready for a quick reproduction

**Comment by rskapadiadev** [comment]
**This below snippet fixed my issue**

```
export const getFBRefreshToken = () => {
  return messaging()
    .deleteToken(undefined, '*')
    .then(async () => {
      const newToken = await messaging().getToken();
      return newToken;
    });
};
```

**Comment by mikehardy** [comment]
Firebase-ios-sdk 8.6.0 included in react-native-firebase 12.7.0 should have another underlying fix for this so that deleteToken really works?
